File: ruby/robot_name.rb
class Robot
  attr_accessor :name

  def initialize
    reset
  end

  def reset
    @name = NameMaker.new_name
  end

end


class NameMaker

  def self.new_name
    @names ||= []
    @names << make_name
    @names.last
  end

  def self.make_name
    puts @names
    ("A".."Z").to_a.sample(2).join + (100...999).to_a.sample.to_s
  end
end

# doesn't pass final test...

File: js/robot-name.js
NameCollector=function(){
  this.collection=[];
};

names = new NameCollector();

Robot = function() {
  this.name = makeName();
  this.reset = reset;

  function makeName() {
    var newName = randomLetters() + randomNumbers()
    if (names.collection.indexOf(newName)!==-1) {
      return makeName();
    }
    else{
      names.collection.push(newName);
      return newName
    }
  };

  function reset(){
    this.name=makeName();
  };

  function randomLetters() {
    var alpha = ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L",
      "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"]
    random1 = alpha[Math.floor(Math.random()*26)];
    random2 = alpha[Math.floor(Math.random()*26)];
    return random1+random2
  };

  function randomNumbers() {
    return Math.random().toString().substring(2,5)
  };
};

module.exports = Robot;
